source/gloperate/source/painter/OrthographicProjectionCapability.cpp

```diff
@@ -0,0 +1,156 @@
+
+#include <gloperate/painter/OrthographicProjectionCapability.h>
+
+#include <glm/glm.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+#include <glm/gtc/constants.hpp>
+
+#include <gloperate/painter/AbstractViewportCapability.h>
+
+
+namespace gloperate {
+
+
+OrthographicProjectionCapability::OrthographicProjectionCapability(AbstractViewportCapability * viewportCapability)
+: AbstractOrthographicProjectionCapability(viewportCapability)
+, m_dirty(false)
+, m_height(1.0f)
+, m_aspect(1.0f)
+, m_zNear(0.1f)
+, m_zFar(64.0f)
+{
+}
+
+OrthographicProjectionCapability::~OrthographicProjectionCapability()
+{
+}
+
+float OrthographicProjectionCapability::zNear() const
+{
+    return m_zNear;
+}
+
+void OrthographicProjectionCapability::setZNear(const float zNear)
+{
+    if (std::abs(zNear - m_zNear) < std::numeric_limits<float>::epsilon())
+        return;
+
+    m_zNear = zNear;
+    assert(m_zNear > 0.0);
+
+    dirty();
+    setChanged(true);
+}
+
+float OrthographicProjectionCapability::zFar() const
+{
+    return m_zFar;
+}
+
+void OrthographicProjectionCapability::setZFar(const float zFar)
+{
+    if (std::abs(zFar - m_zFar) < std::numeric_limits<float>::epsilon())
+        return;
+
+    m_zFar = zFar;
+    assert(m_zFar > m_zNear);
+
+    dirty();
+    setChanged(true);
+}
+
+float OrthographicProjectionCapability::height() const
+{
+    return m_height;
+}
+
+void OrthographicProjectionCapability::setHeight(const float height)
+{
+    if (std::abs(height - m_height) < std::numeric_limits<float>::epsilon())
+        return;
+
+    m_height = height;
+    assert(m_height > 0.0);
+
+    dirty();
+    setChanged(true);
+}
+
+float OrthographicProjectionCapability::aspectRatio() const
+{
+    return m_aspect;
+}
+
+const glm::mat4 & OrthographicProjectionCapability::projection() const
+{
+    if (m_dirty)
+        update();
+
+    if (!m_projection.isValid())
+        m_projection.setValue(projectionForAspectRatio(m_aspect));
+
+    return m_projection.value();
+}
+
+const glm::mat4 & OrthographicProjectionCapability::projectionInverted() const
+{
+    if (m_dirty)
+        update();
+
+    if (!m_projectionInverted.isValid())
+        m_projectionInverted.setValue(glm::inverse(projection()));
+
+    return m_projectionInverted.value();
+}
+
+glm::mat4 OrthographicProjectionCapability::projectionForAspectRatio(float ratio) const
+{
+    const auto width = ratio * m_height;
+    return glm::ortho(-width / 2.0f, width / 2.0f, -m_height / 2.0f, m_height / 2.0f, m_zNear, m_zFar);
```

Comment:
minor detail: personally, i favor multiplication over devision (i'm not sure if current compilers can actually do optimize this ... so i would prefer * 0.5f ... )